### DISCUSSION TITLE
New Aztec detector, feedback welcome

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#402] [Announcements]
I just pushed a new implementation of the Aztec detector code: 192d7e742e6cadac40dba6bbcea87f84da894804

This added around 140 additional successful tests but might have broken others. I'd be interested to know from @vkrause, @gitlost  and @duanqn if they can find issues with their respective test sets.

I'm also interested in getting new tests that still fail. Big symbols (like the ones used by Deutsche Bahn) still don't work well and also the multi-symbol support is still missing. Any other interesting input?

EDIT: multi-symbol support was added in 3430495d06949732f9e40e959977e7c4b5e043d9.

**Comment by duanqn** [comment]
Did a quick test over https://github.com/zxing/zxing/tree/master/core/src/test/resources/blackbox
There is only one Aztec image that still cannot be detected: https://github.com/zxing/zxing/blob/master/core/src/test/resources/blackbox/aztec-2/20.png

**Reply by duanqn** [reply]
Tested on commit 047143277e7810600292bc8a52ddb203248c7e8f. Downscale factor = 2; Downscale threshold = 150; Other parameters are set as default.

**Reply by axxel** [reply]
Thanks for the info but I am of course well aware of that as this is the test-set that I'm working with all the time. I was interested in tests with images outside of that set.

**Reply by duanqn** [reply]
Oh. In that case I don't have more data right now 👀

**Comment by vkrause** [comment]
Ran our test suite with this, looks very good :) No performance or detection regressions, the only deviations were due to codes now being detected that previously failed (e.g. many SNCF Ouigo tickets).

I noticed an unrelated issue though, building with an external libzueci didn't work out of the box, while automatically fetching it was no problem. While that doesn't matter for testing, Linux distributions usually don't like (ie. block) network operations during package builds, so that's something worth investigating before the next release.

**Comment by elejke** [comment]
Hi @axxel!

Hven't you thought about fitting a regression on as much corners as you can get from the bullseye? It looks like if we can get for example {center_point, 2x  inner corner for first black square (along each diagonal) , 2x inner corner for second white square, etc.} we will have 11 or 13 points for each of two diagonals, after that we can do re-projection of points to this regressed diagonal lines and try to estimate "real module size" along diagonals more accurately and after that we can choose some points on this lines (according to the estimated "real" module size) to do the perspective transform.

As for now it looks like we are just getting the avg of two corners (a little bit better for 7-module sized bullseye), but it is still not enough to get accurate transform finally
![Screenshot from 2023-02-19 00-32-11](https://user-images.githubusercontent.com/2506482/219900992-003b11a9-d4b7-43f8-9b04-baee06b846be.png)

Additionally - if you want to get more real-life examples, you can check for [this](https://github.com/zxing-cpp/zxing-cpp/files/10775261/aztec_samples.zip) (collected from the internet)

**Reply by axxel** [reply]
> Hven't you thought about fitting a regression on as much corners as you can get from the bullseye?

No I have not, since I don't believe there is much benefit in doing so. I'm afraid that real-world images like that will only work if the sampling grid is sub-divided much like I just recently did for the QRCode symbology. Unfortunately, with Aztec this is not as straight forward since there are easily detectable alignment patterns missing. One would have to move along the timing patterns and find points to sub-divide the grid this way.

I'm happy to become corrected if my above assumption/opinion is wrong. If you have a PR that improves the success rate, feel free to let me know.

I'm currently focused on QRCode improvements as there is some sponsorship involved (hint, hint ;)).

**Reply by axxel** [reply]
Addendum: I had a look at your samples and saw that most suffer from one particular issue and that is low resolution. In that case I figured it might be worth averaging more pixels to reduce the quantization errors and tested the combination of the 3rd and 2nd white ring. It turns out this actually performs worse. This is not super surprising, though, given that the precision from the 3rd ring can generally be expected to be higher than that of the 2nd, so blending both can easily lower the precision compared to only using the 3rd ring.

Again: if you find something else, let me know.

**Reply by elejke** [reply]
Thank you for your research! 
Yes, I will try to not do the averaging but some kind of line fitting I think and will get back with insides.

**Reply by axxel** [reply]
I was talking about averaging points that have been found by intersecting lines, which in turn have been fitted to points. I'm looking forward to your research results.
